juju/juju
# Ensure Consistent Order of Addresses Returned by Uniter NetworkInfo

## Description of change

This patch ensures that calls to the uniter `NetworkInfo` API method return consistently ordered results, and in the case of AWS, favour non-fan addresses as the reported ingress address.

Some binding/relation networking concerns previously residing in state have been relocated to the uniter API server in a new type dedicated to handling unit networking concerns.

In the state package, `machine.GetNetworkInfoForSpaces` now recieves `network.SpaceInfos` instead of a set of space name strings. This was necessary due to subnet spaces now being identified by ID instead of name.

Some accompanying modularisation and efficiency gains are included in the patch, but this particular area of the code is sufficiently nasty as to make an ideal representation of the logic untenable at this point. I have added comments where further performance may be sought, and I believe the new form will make such investigations easier.

## QA steps

- Bootstrap to AWS.
- `juju add-space new-space 172.31.0.0/20`.
- `juju deploy keystone --bind "internal=new-space --constraints spaces=new-space"`.
- Once deployment is complete, `juju run --unit keystone/0 "network-get --ingress-address internal"` multiple times.
- The non-fan address should always be returned.

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1844148
